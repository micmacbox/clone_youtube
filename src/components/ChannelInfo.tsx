import React, { FC } from "react";
import { useYoutubeApi } from "../context/YoutubeApiContext";
import { useQuery } from "@tanstack/react-query";

type Props = {
  id: string;
  name: string;
};

const ChannelInfo: FC<Props> = ({ id, name }) => {
  const youtube = useYoutubeApi();
  const { data: url } = useQuery({
    queryKey: ["channelImage", id],
    queryFn: () => youtube.channelImageURL(id),
    staleTime: 1000 * 60 * 5,
  });
  return (
    <div className="flex my-4 mb-8 items-center">
      {url && <img className="w-10 h-10 rounded-full" src={url} alt={name} />}
      <p className="text-lg font-medium ml-2">{name}</p>
    </div>
  );
};

export default ChannelInfo;
