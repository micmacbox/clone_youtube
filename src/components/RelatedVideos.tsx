import { useYoutubeApi } from "../context";
import { useQuery } from "@tanstack/react-query";
import { FC } from "react";
import VideoCard from "./VideoCard";
import { VideoItem } from "../@types/video";
import { VideoType } from "../constants";

type Props = {
  id: string;
};

const RelatedVideos: FC<Props> = ({ id }) => {
  const youtube = useYoutubeApi();
  const {
    isLoading,
    error,
    data: videos,
  } = useQuery({
    queryKey: ["channel", id],
    queryFn: () => youtube.channelVideos(id),
    staleTime: 1000 * 60 * 5,
  });
  return (
    <>
      {isLoading && <p>loading...</p>}
      {error && <p>error occured ðŸ˜°</p>}
      {videos && (
        <ul>
          {videos.map((video: VideoItem) => (
            <VideoCard
              key={video.id}
              video={video}
              type={VideoType.relatedVideos}
            />
          ))}
        </ul>
      )}
    </>
  );
};

export default RelatedVideos;
