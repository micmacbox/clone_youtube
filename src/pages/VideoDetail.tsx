import React, { FC } from "react";
import { useLocation } from "react-router-dom";
import ChannelInfo from "../components/ChannelInfo";
import RelatedVideos from "../components/RelatedVideos";

type Props = {};

const VideoDetail: FC<Props> = ({}) => {
  const {
    state: {
      video: {
        id,
        snippet: { title, channelId, channelTitle, description },
      },
    },
  } = useLocation();

  return (
    <section className="flex flex-col lg:flex-row">
      <article className="basis-4/6">
        <iframe
          title="player"
          id="player"
          width="100%"
          height="640"
          src={`http://www.youtube.com/embed/${id}`}
        />
        <div className="p-8">
          <h2 className="text-xl font-bold">{title}</h2>
          <ChannelInfo id={channelId} name={channelTitle} />
          <pre className="whitespace-pre-wrap">{description}</pre>
        </div>
      </article>
      <section className="basis-2/6">
        <RelatedVideos id={channelId} />
      </section>
    </section>
  );
};

export default VideoDetail;
