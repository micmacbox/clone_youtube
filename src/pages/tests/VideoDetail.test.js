import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { withRouter } from "../../test/utils";
import { Route } from "react-router-dom";
import ChannelInfo from "../../components/ChannelInfo";
import RelatedVideos from "../../components/RelatedVideos";
import VideoDetail from "../VideoDetail";
import { fakeVideo } from "../../test";

jest.mock("../../components/ChannelInfo");
jest.mock("../../components/RelatedVideos");

describe("VideoDetail", () => {
  afterEach(() => {
    ChannelInfo.mockReset();
    RelatedVideos.mockReset();
  });

  it("renders video item details", () => {
    render(
      withRouter(<Route path="/" element={<VideoDetail />} />, {
        pathname: "/",
        state: { video: fakeVideo },
        key: "fake-key",
      })
    );
  });

  const { title, channelId, channelTitle } = fakeVideo.snippet;
  expect(screen.getByTitle(title)).toBeInTheDocument();
  expect(RelatedVideos.mock.calls[0][0]).toStricrtEqual({ id: fakeVideo.id });
  expect(ChannelInfo.mock.calls[0][0]).toStrictEqual({
    id: channelId,
    name: channelTitle,
  });
});
