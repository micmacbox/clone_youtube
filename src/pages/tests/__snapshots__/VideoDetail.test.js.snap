// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetail renders correctly 1`] = `
<div>
  NotFound
</div>
`;
